staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go

```diff
@@ -21,31 +21,54 @@ import (
 	"fmt"
 	"net/http"
 	"strings"
+	"sync"
 
 	"k8s.io/apiserver/pkg/authentication/user"
 	"k8s.io/apiserver/pkg/endpoints/request"
-	"k8s.io/kube-openapi/pkg/cached"
 	"k8s.io/kube-openapi/pkg/validation/spec"
 )
 
+type CacheableDownloader interface {
+	UpdateHandler(http.Handler)
+	Get() (*spec.Swagger, string, error)
+}
+
 // cacheableDownloader is a downloader that will always return the data
 // and the etag.
 type cacheableDownloader struct {
+	name       string
 	downloader *Downloader
 	handler    http.Handler
 	etag       string
 	spec       *spec.Swagger
+	mutex      sync.RWMutex
 }
 
 // NewCacheableDownloader creates a downloader that also returns the etag, making it useful to use as a cached dependency.
-func NewCacheableDownloader(downloader *Downloader, handler http.Handler) cached.Value[*spec.Swagger] {
+func NewCacheableDownloader(apiServiceName string, downloader *Downloader, handler http.Handler) CacheableDownloader {
 	return &cacheableDownloader{
+		name:       apiServiceName,
 		downloader: downloader,
 		handler:    handler,
 	}
 }
+func (d *cacheableDownloader) UpdateHandler(handler http.Handler) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	d.handler = handler
+}
 
 func (d *cacheableDownloader) Get() (*spec.Swagger, string, error) {
```

Comment:
Get is long-running, right? the mutex addition makes that block UpdateHandler, which now blocks specAggregator#AddUpdateAPIService

That means a hang fetching one service's openapi can block processing of other services, right?

Reply:
hmm that's true. What if we have the mutex only be required when we fetch the handler instead of the entire get operation?

Reply:
is the only thing we need to guard the handler, or is the etag also concurrently modifiable? (can Get be called from multiple threads?)

Reply:
We currently don't call Get() from multiple threads but the defacto `cached.Value` interface that we follow is supposed to be thread safe (https://github.com/kubernetes/kube-openapi/blob/master/pkg/cached/cache.go#L17-L22). 

@apelisse I think this ties into what you mentioned earlier that since we don't evaluate aggregated apiservices lazily, maybe we shouldn't follow cached library interface?

Reply:
leaving this open before merge just to get @apelisse's feedback

Reply:
If you don't then you'll have to deal with the etag manually, I'm not sure it would better.